external/llvm-project/mlir/lib/Dialect/LLVMIR/Transforms/SoftwareBf16.cpp

```diff
@@ -0,0 +1,364 @@
+      bitcastType = srcShaped.clone(bitcastType);
+
+    Type i16 = rewriter.getIntegerType(16);
+    if (auto destShaped = destType.dyn_cast<ShapedType>()) {
+      if (destShaped.getElementType() != i16)
+        return failure();
+    } else if (destType != i16)
+      return failure();
+
+    // a = bitcast f32 value to i32
+    // b = (a + 32767) << 16
+    // c = ((a << 16) & 1)
+    // d = b + c
+    // truncate (d << 16) to i16 and return this i16
+    Value bitcastop =
+        rewriter.create<LLVM::BitcastOp>(loc, bitcastType, op.getArg());
+    Value constantSixteen = getLlvmI32Const(loc, rewriter, bitcastType, 16);
+    Value shiftValue = rewriter.create<LLVM::LShrOp>(
+        loc, bitcastType, bitcastop, constantSixteen);
+
+    Value constantOne = getLlvmI32Const(loc, rewriter, bitcastType, 1);
+    Value andValue = rewriter.create<LLVM::AndOp>(loc, shiftValue, constantOne);
+
+    Value constantBig = getLlvmI32Const(loc, rewriter, bitcastType, 32767);
+    Value addBigValue =
+        rewriter.create<LLVM::AddOp>(loc, bitcastop, constantBig);
+    Value addValue = rewriter.create<LLVM::AddOp>(loc, andValue, addBigValue);
+
+    Value shiftBeforeTruncValue = rewriter.create<LLVM::LShrOp>(
+        loc, bitcastType, addValue, constantSixteen);
+    Value truncValue =
+        rewriter.create<LLVM::TruncOp>(loc, destType, shiftBeforeTruncValue);
+    rewriter.replaceOp(op.getOperation(), {truncValue});
+
+    return success();
+  }
+};
+
+void replaceBF16InOp(Operation *op, LLVMTypeConverter &converter);
+
+void replaceBF16InBlock(Block &block, LLVMTypeConverter &converter) {
+  for (Operation &op : block.getOperations())
+    replaceBF16InOp(&op, converter);
+}
+
+void replaceBF16InRegion(Region &region, LLVMTypeConverter &converter) {
+  for (Block &block : region.getBlocks())
+    replaceBF16InBlock(block, converter);
+}
+
+void replaceBF16InOp(Operation *op, LLVMTypeConverter &converter) {
+  if (auto func = dyn_cast<LLVM::LLVMFuncOp>(op)) {
+    auto funcType = func.getFunctionType();
+    func.setType(converter.convertType(funcType));
+    for (auto arg : func.getArguments())
+      arg.setType(converter.convertType(arg.getType()));
+  } else {
+    for (unsigned idx = 0; idx < op->getNumOperands(); idx++) {
+      auto type = converter.convertType(op->getOperand(idx).getType());
+      op->getOperand(idx).setType(type);
+    }
+    for (unsigned idx = 0; idx < op->getNumResults(); idx++) {
+      auto type = converter.convertType(op->getResult(idx).getType());
+      op->getResult(0).setType(type);
+    }
+  }
+  for (Region &region : op->getRegions())
+    replaceBF16InRegion(region, converter);
+  return;
+}
+
+} // namespace LLVM
+} // namespace mlir
+
+void mlir::LLVM::populateBF16ToLLVMConversionPatterns(
+    LLVMTypeConverter &converter, RewritePatternSet &patterns) {
+  MLIRContext *ctx = converter.getDialect()->getContext();
+  // AMD GPUs don't have a backend that understands bfloat, even though LLVM's
```

Comment:
Make comment slightly more generic - maybe "Some LLVM targets, such as"

Reply:
This here comment can probably disappear, since the pass documentation explains what it does.